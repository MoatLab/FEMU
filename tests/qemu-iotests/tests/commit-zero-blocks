#!/usr/bin/env bash
# group: rw quick
#
# Test for commit of discarded blocks
#
# This tests committing a live snapshot where some of the blocks that
# are present in the base image are discarded in the intermediate image.
# This intends to check that these blocks are also discarded in the base
# image after the commit.
#
# Copyright (C) 2024 Vincent Vanlaer.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# creator
owner=libvirt-e6954efa@volkihar.be

seq=`basename $0`
echo "QA output created by $seq"

status=1	# failure is the default!

_cleanup()
{
    _cleanup_qemu
    _rm_test_img "${TEST_IMG}.base"
    _rm_test_img "${TEST_IMG}.mid"
    _cleanup_test_img
}
trap "_cleanup; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
cd ..
. ./common.rc
. ./common.filter
. ./common.qemu

_supported_fmt qcow2
_supported_proto file

size="1M"

TEST_IMG="$TEST_IMG.base" _make_test_img $size
TEST_IMG="$TEST_IMG.mid" _make_test_img -b "$TEST_IMG.base" -F $IMGFMT $size
_make_test_img -b "${TEST_IMG}.mid" -F $IMGFMT $size

$QEMU_IO -c "write -P 0x01 64k 128k" "$TEST_IMG.base" | _filter_qemu_io
$QEMU_IO -c "discard 64k 64k" "$TEST_IMG.mid" | _filter_qemu_io

echo
echo "=== Base image info before commit ==="
TEST_IMG="${TEST_IMG}.base" _img_info | _filter_img_info
$QEMU_IMG map --output=json "$TEST_IMG.base" | _filter_qemu_img_map

echo
echo "=== Middle image info before commit ==="
TEST_IMG="${TEST_IMG}.mid" _img_info | _filter_img_info
$QEMU_IMG map --output=json "$TEST_IMG.mid" | _filter_qemu_img_map

echo
echo === Running QEMU Live Commit Test ===
echo

qemu_comm_method="qmp"
_launch_qemu -drive file="${TEST_IMG}",if=virtio,id=test
h=$QEMU_HANDLE

_send_qemu_cmd $h "{ 'execute': 'qmp_capabilities' }" "return"

_send_qemu_cmd $h "{ 'execute': 'block-commit',
                                 'arguments': { 'device': 'test',
                                 'top': '"${TEST_IMG}.mid"',
                                 'base': '"${TEST_IMG}.base"'} }" '"status": "null"'

_cleanup_qemu

echo
echo "=== Base image info after commit ==="
TEST_IMG="${TEST_IMG}.base" _img_info | _filter_img_info
$QEMU_IMG map --output=json "$TEST_IMG.base" | _filter_qemu_img_map

# success, all done
echo "*** done"
rm -f $seq.full
status=0
