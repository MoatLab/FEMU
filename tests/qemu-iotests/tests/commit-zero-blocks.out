QA output created by commit-zero-blocks
Formatting 'TEST_DIR/t.IMGFMT.base', fmt=IMGFMT size=1048576
Formatting 'TEST_DIR/t.IMGFMT.mid', fmt=IMGFMT size=1048576 backing_file=TEST_DIR/t.IMGFMT.base backing_fmt=IMGFMT
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=1048576 backing_file=TEST_DIR/t.IMGFMT.mid backing_fmt=IMGFMT
wrote 131072/131072 bytes at offset 65536
128 KiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
discard 65536/65536 bytes at offset 65536
64 KiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

=== Base image info before commit ===
image: TEST_DIR/t.IMGFMT.base
file format: IMGFMT
virtual size: 1 MiB (1048576 bytes)
[{ "start": 0, "length": 65536, "depth": 0, "present": false, "zero": true, "data": false, "compressed": false},
{ "start": 65536, "length": 131072, "depth": 0, "present": true, "zero": false, "data": true, "compressed": false, "offset": OFFSET},
{ "start": 196608, "length": 851968, "depth": 0, "present": false, "zero": true, "data": false, "compressed": false}]

=== Middle image info before commit ===
image: TEST_DIR/t.IMGFMT.mid
file format: IMGFMT
virtual size: 1 MiB (1048576 bytes)
backing file: TEST_DIR/t.IMGFMT.base
backing file format: IMGFMT
[{ "start": 0, "length": 65536, "depth": 1, "present": false, "zero": true, "data": false, "compressed": false},
{ "start": 65536, "length": 65536, "depth": 0, "present": true, "zero": true, "data": false, "compressed": false},
{ "start": 131072, "length": 65536, "depth": 1, "present": true, "zero": false, "data": true, "compressed": false, "offset": OFFSET},
{ "start": 196608, "length": 851968, "depth": 1, "present": false, "zero": true, "data": false, "compressed": false}]

=== Running QEMU Live Commit Test ===

{ 'execute': 'qmp_capabilities' }
{"return": {}}
{ 'execute': 'block-commit',
                                 'arguments': { 'device': 'test',
                                 'top': 'TEST_DIR/t.IMGFMT.mid',
                                 'base': 'TEST_DIR/t.IMGFMT.base'} }
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "JOB_STATUS_CHANGE", "data": {"status": "created", "id": "test"}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "JOB_STATUS_CHANGE", "data": {"status": "running", "id": "test"}}
{"return": {}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "JOB_STATUS_CHANGE", "data": {"status": "waiting", "id": "test"}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "JOB_STATUS_CHANGE", "data": {"status": "pending", "id": "test"}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "BLOCK_JOB_COMPLETED", "data": {"device": "test", "len": 1048576, "offset": 1048576, "speed": 0, "type": "commit"}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "JOB_STATUS_CHANGE", "data": {"status": "concluded", "id": "test"}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "JOB_STATUS_CHANGE", "data": {"status": "null", "id": "test"}}

=== Base image info after commit ===
image: TEST_DIR/t.IMGFMT.base
file format: IMGFMT
virtual size: 1 MiB (1048576 bytes)
[{ "start": 0, "length": 65536, "depth": 0, "present": false, "zero": true, "data": false, "compressed": false},
{ "start": 65536, "length": 65536, "depth": 0, "present": true, "zero": true, "data": false, "compressed": false},
{ "start": 131072, "length": 65536, "depth": 0, "present": true, "zero": false, "data": true, "compressed": false, "offset": OFFSET},
{ "start": 196608, "length": 851968, "depth": 0, "present": false, "zero": true, "data": false, "compressed": false}]
*** done
